File: solutions/cpp/lasagna-master/1/lasagna_master.cpp
#include "lasagna_master.h"

namespace lasagna_master {

int preparationTime(const std::vector<std::string>& layers, int avgPrepTime) {
    return layers.size() * avgPrepTime;
}

amount quantities(const std::vector<std::string>& items) {
    int NOODLE_MASS_G{50};
    double SAUCE_VOLUME_L{0.2};
    amount result{};
    
    for (std::string item : items) {
        if (item == "noodles") {
            result.noodles += NOODLE_MASS_G;
        } else if (item == "sauce") {
            result.sauce += SAUCE_VOLUME_L;
        }
    }
    return result;
}

void addSecretIngredient(std::vector<std::string>& mine, const std::vector<std::string>& theirs) {
    if (mine.size() > 0 && theirs.size() > 0) {
        mine[mine.size() - 1] = theirs[theirs.size() - 1];
    }
}

void addSecretIngredient(std::vector<std::string>& mine, std::string theirs) {
    if (mine.size() > 0) {
        mine[mine.size() - 1] = theirs;
    }
}

std::vector<double> scaleRecipe(std::vector<double> twoPortionQuantity, int desiredQuantities) {
    double multiplier{desiredQuantities * 0.5};
    std::vector<double> result{};

    for (double item : twoPortionQuantity) {
        result.push_back(item * multiplier);
    }
    
    return result;
}
    
}  // namespace lasagna_master

File: solutions/go/lasagna-master/1/lasagna_master.go
package lasagnamaster

func PreparationTime(layers []string, timePerLayer int) int {
	numLayers := len(layers)
	if timePerLayer == 0 {
		timePerLayer = 2
	}
	return numLayers * timePerLayer
}

func Quantities(layers []string) (int, float64) {
	noodleLayerGrams := 50
	sauceLayerLiters := 0.2
	noodlesTotMass := 0
	sauceTotVolume := 0.0

	for _, v := range layers {
		if v == "noodles" {
			noodlesTotMass += noodleLayerGrams
		}

		if v == "sauce" {
			sauceTotVolume += sauceLayerLiters
		}
	}
	return noodlesTotMass, sauceTotVolume

}

func AddSecretIngredient(friendRecipe []string, ownRecipe []string) {
	specialIngredient := friendRecipe[len(friendRecipe)-1]
	ownRecipe[len(ownRecipe)-1] = specialIngredient
}

func ScaleRecipe(quantities []float64, portions int) []float64 {
	resultQuantities := make([]float64, 0, len(quantities))
	for _, v := range quantities {
		resultQuantities = append(resultQuantities, v*(float64(portions)/2.0))
	}
	return resultQuantities
}

// Your first steps could be to read through the tasks, and create
// these functions with their correct parameter lists and return types.
// The function body only needs to contain `panic("")`.
//
// This will make the tests compile, but they will fail.
// You can then implement the function logic one by one and see
// an increasing number of tests passing as you implement more
// functionality.
